### ISSUE TITLE
Support for RowN.mapping(Function<? super Object[], ? extends U>)

### ISSUE THREAD

**Issue by alf** [jOOQ/jOOQ#12515]
### Use case:

We have situations where we work on groups of fields where the group is of unknown size. The current use case is composite keys, but I can see this as useful for other groups as well.

By wrapping the fields in a `row` we can attach a mapping function to the resulting selectfield at query build time, something that is very powerful when building dynamic reusable queries. 

Unfortunately the RowN interface does not provide the `mapping` method, thus we have to switch over the group size and create a lot of boilerplate/duplicate code.

### Possible solution you'd like to see:

Being able to attach a mapping on a RowN. This mapping could provide us with a Record that can be used together with the original group of fields to get the values and instantiate the user class we want.

The parameters provided to the mapping-method could include the group of fields as well as the record, but this isn’t strictly necessary. 

### Possible workarounds:

Switch over the group size and create the specific methods needed.

**Comment by alf** [comment]
I understand the issues with arity and suspected that this feature might be wishful thinking. I figured raising it might be worth while even so. I'm not too unhappy with the switch-based approach, it just feels a bit off. :)

I looked at using embedded keys, but ran into issues since we make extensive use of natural keys and need to access the columns individually as well. I found no way of doing that. Didn't think of raising this as an issue to be honest.

Here's a snippet of what I'm currently doing:

```
public static SelectField<String> getId(Table<?> table, Key<?> key) {
        var fs = keyFields(table, key);

        switch (fs.size()) {
            case 1:
                return createRow(fs.get(0));
            case 2:
                return createRow(fs.get(0), fs.get(1));
            case 3:
                return createRow(fs.get(0), fs.get(1), fs.get(2));
            case 4:
                return createRow(fs.get(0), fs.get(1), fs.get(2), fs.get(3));
            case 5:
                return createRow(fs.get(0), fs.get(1), fs.get(2), fs.get(3), fs.get(4));
            default:
                throw new IllegalArgumentException("getId only supports keys with 5 columns, we must add more");
        }
    }

    private static <T1, T2, T3, T4, T5> SelectField<String> createRow(Field<T1> f1, Field<T2> f2, Field<T3> f3, Field<T4> f4, Field<T5> f5) {
        return row(f1, f2, f3, f4, f5).mapping(String.class, (v1, v2, v3, v4, v5) -> csv(toStr(f1, v1), toStr(f2, v2), toStr(f3, v3), toStr(f4, v4), toStr(f5, v5)));
    }

[snip]
```

**Comment by lukaseder** [comment]
I just needed this myself, recently, and was probably as surprised as you were, when it wasn't there. The most obvious choice of method signature would be:

```java
<U> SelectField<U> mapping(Function<? super Object[], ? extends U> function);
<U> SelectField<U> mapping(Class<U> uType, Function<? super Object[], ? extends U> function);
```

An alternative would be to apply `Collection<?>` to the argument function, as we always overload `T|Collection<T>` throughout the API, including the  `DSL::row` constructors. I have a feeling that `Object[]` is more consistent with the rest of jOOQ, even if it's a bit less convenient.

**Comment by lukaseder** [comment]
Implemented in jOOQ 3.17.0. Thanks again for your feature request.
